plugins/speculation-rules/tests/test-speculation-rules-settings.php

```diff
@@ -122,4 +122,156 @@ public function test_plsr_add_settings_action_link(): void {
+		$sanitized = plsr_sanitize_setting( $input );
+		$this->assertEquals( plsr_get_setting_default(), $sanitized );
+	}
+
+	/**
+	 * @covers ::plsr_add_setting_ui
+	 */
+	public function test_add_setting_ui(): void {
+		do_action( 'load-options-reading.php' );// phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
+
+		// Check if the settings section has been added.
+		global $wp_settings_sections;
+		$this->assertArrayHasKey( 'reading', $wp_settings_sections );
+		$this->assertArrayHasKey( 'plsr_speculation_rules', $wp_settings_sections['reading'] );
+
+		// Check the output of the callback function for the section.
+		ob_start();
+		call_user_func( $wp_settings_sections['reading']['plsr_speculation_rules']['callback'] );
+		$output = ob_get_clean();
+		$this->assertStringContainsString( 'This section allows you to control how URLs that your users navigate to are speculatively loaded to improve performance.', $output );
+	}
+
+	/**
+	 * Data provider for testing plsr_render_settings_field.
+	 *
+	 * @return array<string, array<mixed>> Data for testing settings fields.
+	 */
+	public function settingsFieldProvider(): array {
+		return array(
+			'mode'      => array(
+				array(
+					'field'       => 'mode',
+					'title'       => 'Speculation Mode',
+					'description' => 'Prerendering will lead to faster load times than prefetching.',
+				),
+				array(
+					'mode'      => 'prefetch',
+					'eagerness' => 'moderate',
+				),
+				'name="plsr_speculation_rules[mode]"',
+				'value="prefetch"',
+				'Prerendering will lead to faster load times than prefetching.',
+			),
+			'eagerness' => array(
+				array(
+					'field'       => 'eagerness',
+					'title'       => 'Eagerness',
+					'description' => 'The eagerness setting defines the heuristics based on which the loading is triggered.',
+				),
+				array(
+					'mode'      => 'prefetch',
+					'eagerness' => 'moderate',
+				),
+				'name="plsr_speculation_rules[eagerness]"',
+				'value="moderate"',
+				'The eagerness setting defines the heuristics based on which the loading is triggered.',
+			),
+		);
+	}
+
+	/**
+	 * Test rendering of settings fields using data provider.
+	 *
+	 * @dataProvider settingsFieldProvider
+	 * @param array<mixed> $args Arguments for the settings field.
+	 * @param array<mixed> $stored_settings Stored settings values.
+	 * @param string       $name_check HTML name attribute check.
+	 * @param string       $value_check HTML value attribute check.
+	 * @param string       $description_check Description check.
+	 */
+	public function test_render_settings_field( array $args, array $stored_settings, string $name_check, string $value_check, string $description_check ): void {
+		// Simulate getting stored settings.
+		update_option( 'plsr_speculation_rules', $stored_settings );
+
+		// Capture the output of the settings field rendering.
+		ob_start();
+		plsr_render_settings_field( $args );
+		$output = ob_get_clean();
```

Comment:
There is the [`get_echo()`](https://github.com/WordPress/wordpress-develop/blob/7d10dd7b0fde2a782395887c2d66439481440f9b/tests/phpunit/includes/utils.php#L432-L436) helper function available in WP unit tests which will reduce some of this code:

```suggestion
		$output = get_echo( 'plsr_render_settings_field', array( $args ) );
```